python/lsst/ts/m2/tcp_client.py

```diff
@@ -0,0 +1,371 @@
+
+        self.reader = None
+        self.writer = None
+
+        self.max_n_bytes = int(max_n_bytes)
+
+        # Unique ID
+        self._uniq_id = self._uniq_id_gen()
+
+        self.queue = asyncio.Queue(maxsize=int(maxsize_queue))
+
+        # Monitor loop task (asyncio.Task)
+        self._monitor_loop_task = None
+
+        # Monitor loop task is done or not (asyncio.Future)
+        self._monitor_loop_task_done = None
+
+    def _uniq_id_gen(self):
+        """Unique ID generator.
+
+        Yields
+        ------
+        `Generator`
+            Unique Id generator object.
+        """
+
+        value = 0
+        while True:
+            yield value
+            value += 1
+
+    def get_uniq_id(self):
+        """Get the unique ID.
+
+        Returns
+        -------
+        `int`
+            Unique ID.
+        """
+
+        return next(self._uniq_id)
+
+    async def connect(self, connect_retry_interval=1):
+        """Connect to the server.
+
+        Parameters
+        ----------
+        connect_retry_interval : `float`, optional
+            Connection retry interval in second. (default is 1)
+
+        Notes
+        -----
+        This will wait forever for a connection.
+        """
+
+        self.log.info("Open the connection.")
+
+        while not self.is_connected():
+
+            try:
+                self.reader, self.writer = await asyncio.open_connection(
+                    host=self.host, port=self.port
+                )
+
+            except ConnectionRefusedError:
+                await asyncio.sleep(connect_retry_interval)
+
+        self.log.info("Connection is on.")
+
+    def is_connected(self):
+        """Is connected with the server or not.
+
+        Returns
+        -------
+        bool
+            True if is connected. Else, False.
+        """
+
+        return not (
+            self.reader is None
+            or self.writer is None
+            or self.reader.at_eof()
+            or self.writer.is_closing()
+        )
+
+    async def write(self, topic_type, topic, topic_details=None, comp_name=None):
+        """Write the message.
+
+        Parameters
+        ----------
+        topic_type : `TopicType`
+            Topic type.
+        topic : `str`
+            Topic name.
+        topic_details : `dict` or None, optional
+            Topic details. (the default is None)
+        comp_name : `str` or None, optional
+            Specific component name used in the event or telemetry. (the
+            default is None)
+
+        Raises
+        ------
+        RuntimeError
+            There is no TCP/IP connection.
+        ValueError
+            The topic type is not supported.
+        """
+
+        if not self.is_connected():
+            raise RuntimeError("There is no TCP/IP connection.")
+
+        if topic_details is None:
+            topic_details_with_header = dict()
+        else:
+            topic_details_with_header = copy.copy(topic_details)
```

Comment:
The reason of this is to avoid the duplication of code if the `topic_details=None`. This is allowed for the event such as `heartbeat` or `inPosition`. Only the name is needed at this time. Otherwise, I will need to copy the following code in` _add_*_header()` method:
```
if topic_details is None:
    topic_details_with_header = dict()
else:
    topic_details_with_header = copy.copy(topic_details)
```

BTW, they are the internal methods and I believe the operation of object is not a concern here. If they are the public (external) use methods, I agree with you that I should use the `return` instead.

Thanks!

Reply:
You could do this with one inline call;

```
topic_details_with_header = copy.copy(topic_details) if topic_details is not None else dict()
```

This is actually faster than the `if/else" statement. 

If you insist in keeping this design, please add a comment in the code that the `_add_*_headers` are modifying the input dictionary.

Reply:
Added the comments in the commit: a9df069. Thanks!